calculator/calculator.py

```diff
@@ -1,15 +1,57 @@
 import re
-
-
-class Calculator:
-
-    @classmethod
-    def calculate(cls, expression=str):
-        if type(expression) is not str:
-            raise TypeError("Expression input should be of str type.")
-
-        # TODO
-
-
-
-
+import math
+from typing import Union
+from statistics import mean
+
+
+_NUMBER_REGEX = r'\d+(?:\.\d+)?'
+_NUMBER_NEGATIVE_OPT_REGEX = f'-?{_NUMBER_REGEX}'
+_LEFT_PARENTHESIS_REGEX = r'(?:\()'
+_RIGHT_PARENTHESIS_REGEX = r'(?:\))'
+
+
+def calculate(expression: str) -> Union[int, float]:
+    if type(expression) is not str:
+        raise TypeError("Expression input should be of str type.")
+
+    return _calculate(expression.replace(" ", ""))
+
+def _calculate(expression: str) -> Union[int, float]:
+    if _is_end_of_calculation(expression):
+        return expression
+
+    for regex, handler in _TOKEN_SPECIFICATION.items():
+        match = re.search(regex, expression)
+        if match:
+            solved_step_expression = re.sub(match.re, str(handler(*match.groups())), expression, 1)
+            return _calculate(solved_step_expression)
+
+    raise ValueError(f'Expression: {expression} is not legal!')
+
+def _add_operation_variables(operation: str, is_single_variable: bool = False, is_negative_legal: bool = False) -> str:
+    number_regex = _NUMBER_NEGATIVE_OPT_REGEX if is_negative_legal else _NUMBER_REGEX
+    return f'(?P<NUMBER>{number_regex}){operation}' if is_single_variable else f'(?P<FIRST_NUMBER>{number_regex})' \
+                                                                               f'{operation}' \
+                                                                               f'(?P<SECOND_NUMBER>{number_regex})'
```

Comment:
FIRST_NUMBER and SECOND_NUMBER are redundent

Reply:
I use groups to group each expression into the operator and 2 number or 1, so if its 2 different numbers i cant use the same name for their group, what would you do differently?

Reply:
You don't parse any named group, you use *groups.
It's redundant.